### PR TITLE
Add typing to protocol_muxer

### PR DESCRIPTION
**Pull request by ChihChengLiang**

### CODE REVIEW

**Comment by mhchia** [comment]
Would you mind resolving the CI error? https://travis-ci.com/libp2p/py-libp2p/jobs/223135166

File: libp2p/protocol_muxer/multiselect_client_interface.py [hunk 1]
```diff
@@ -8,7 +12,7 @@ class IMultiselectClient(ABC):
     """
 
     @abstractmethod
-    def select_protocol_or_fail(self, protocol, stream):
+    async def select_protocol_or_fail(self, protocol: TProtocol, stream: IMuxedStream) -> TProtocol:
```
Comment (mhchia) [inline libp2p/protocol_muxer/multiselect_client_interface.py]: ditto

File: libp2p/protocol_muxer/multiselect_client_interface.py [hunk 2]
```diff
@@ -18,7 +22,9 @@ def select_protocol_or_fail(self, protocol, stream):
         """
 
     @abstractmethod
-    def select_one_of(self, protocols, stream):
+    async def select_one_of(
+        self, protocols: Sequence[TProtocol], stream: IMuxedStream
```
Comment (mhchia) [inline libp2p/protocol_muxer/multiselect_client_interface.py]: ditto

File: libp2p/protocol_muxer/multiselect_communicator.py [hunk 3]
```diff
@@ -8,22 +10,24 @@ class MultiselectCommunicator(IMultiselectCommunicator):
     which is necessary for them to work
     """
 
-    def __init__(self, reader_writer):
+    reader_writer: IMuxedStream
```
Comment (mhchia) [inline libp2p/protocol_muxer/multiselect_communicator.py]: ditto

File: libp2p/protocol_muxer/multiselect_communicator.py [hunk 4]
```diff
@@ -8,22 +10,24 @@ class MultiselectCommunicator(IMultiselectCommunicator):
...
+
+    def __init__(self, reader_writer: IMuxedStream) -> None:
```
Comment (mhchia) [inline libp2p/protocol_muxer/multiselect_communicator.py]: ditto

File: libp2p/protocol_muxer/multiselect_muxer_interface.py [hunk 5]
```diff
@@ -8,16 +12,18 @@ class IMultiselectMuxer(ABC):
...
+    def add_handler(self, protocol: TProtocol, handler: StreamHandlerFn) -> None:
         """
         Store the handler with the given protocol
         :param protocol: protocol name
         :param handler: handler function
         """
 
     @abstractmethod
-    def negotiate(self, stream):
+    async def negotiate(self, stream: IMuxedStream) -> Tuple[TProtocol, StreamHandlerFn]:
```
Comment (mhchia) [inline libp2p/protocol_muxer/multiselect_muxer_interface.py]: ditto

File: libp2p/protocol_muxer/multiselect.py [hunk 6]
```diff
@@ -12,18 +18,20 @@ class Multiselect(IMultiselectMuxer):
...
+    def add_handler(self, protocol: TProtocol, handler: StreamHandlerFn) -> None:
         """
         Store the handler with the given protocol
         :param protocol: protocol name
         :param handler: handler function
         """
         self.handlers[protocol] = handler
 
-    async def negotiate(self, stream):
+    async def negotiate(self, stream: IMuxedStream) -> Tuple[TProtocol, StreamHandlerFn]:
```
Comment (mhchia) [inline libp2p/protocol_muxer/multiselect.py]:
```suggestion
    async def negotiate(self, stream: Union[IMuxedStream, IRawConnection]) -> Tuple[TProtocol, StreamHandlerFn]:
```
From the code snippet below, we know that `Multiselect` and `MultiselectClient` are used not only for the streams but connections for selecting protocol for connections(security, and multiplexer). Therefore, the type should include `IRawConnection`. I'm thinking we can add some type like `NegotiableTransport = Union[IMuxedStream, IRawConnection]`.
https://github.com/libp2p/py-libp2p/blob/7477b295082e47f92ae7f03398b4cdb4aef9b248/libp2p/security/security_multistream.py#L94-L99

File: tests/pubsub/test_mcache.py [hunk 7]
```diff
@@ -4,6 +4,8 @@
 
 
 class Msg:
+    __slots__ = ["topicIDs", "seqno", "from_id"]
```
Comment (mhchia) [inline tests/pubsub/test_mcache.py]: May I ask the reason why we need `__slots__` here?

Comment (ChihChengLiang) [reply tests/pubsub/test_mcache.py]: to make flake8 happy "B903".

**Pull request merged**